### ISSUE TITLE
Bug in statistics gathering upon first visit of repository (introduction of /csrf endpoint)

### ISSUE THREAD

**Issue by KevinVdV** [DSpace/DSpace#9236]
**Describe the bug**
The first time when you visit the repository the first POST will fail because of you have an invalid CSRF header. The first POST that ANGULAR makes is a POST to the statistics endpoint leading to certain statistics not being logged.

**To Reproduce**
1. Navigate to the https://demo7.dspace.org/ repository next use a browser tool to clear your cookies / session (I used the following one: https://chromewebstore.google.com/detail/clear-session/maejjihldgmkjlfmgpgoebepjchengka?pli=1)
2. Open a network tab
3. Refresh the page (this will be the first visit to the page as we cleared everything)
4. The POST to the statistics endpoint will result in a "403 Forbidden" with the message: "Access is denied. Invalid CSRF token."
5. Subsequent POSTS either using a page refresh or navigating to other pages will work.

**Expected behavior**
The statistic post should work.

**Proposed solution**
In order to fix this we propose to create a dedicated POST endpoint to generate the CSRF header, this POST will be the first thing that Angular does before anything else to ensure that subsequent requests will have an CSRF header/cookie.

**Comment by tdonohue** [comment]
@KevinVdV : This looks to be the same bug as described in https://github.com/DSpace/dspace-angular/issues/1486  See discussion on that ticket.

**Comment by tdonohue** [comment]
@KevinVdV : Reading the description, I see you recommend a new endpoint.  Keep in mind CSRF protection is **built into Spring / Spring Boot**.  Spring will generate the CSRF token for the first GET request & send it back.  It may not be necessary to create a custom endpoint to do this for us.  We also do not want to bypass or override the automatic protections that Spring provides.

More on Spring's CSRF protection options (we are using the Sychronizer Token Pattern): https://docs.spring.io/spring-security/reference/features/exploits/csrf.html

**Comment by benbosman** [comment]
@tdonohue what happens here is actually a race condition. The first GET request will set the cookie and header. But it takes some time to process both the header and the cookie, and both are not processed at the same time. What happens is that REST will send multiple CSRF tokens (with the asynchronous requests to REST), and the end result is often that the header and the cookie don't match.
The custom endpoint is required because we can't rely on the cookie and header from the GET requests

**Comment by tdonohue** [comment]
@benbosman : I agree that it definitely seems like a race condition.  I just wanted to warn that we don't want to be bypassing or overriding CSRF protection in Spring... I'm not against a new endpoint as long as it aligns with Spring.  I just don't quite understand yet why a new `POST` endpoint solves the race condition better than just waiting on the `GET` to return the new CSRF token. 

The *challenge* that I see here is that all `POST` requests REQUIRE a CSRF token (per Spring rules & our own docs..see link below).  So, to create a `POST` endpoint that generates a CSRF token seems problematic itself... you'd then need a CSRF token to generate a CSRF token.  If you bypass CSRF protection on that new endpoint's `POST` then that'd be bypassing some of Spring's protections.  In other words, it might be better to consider making this a `GET` or `HEAD` style request to grab the CSRF token... that way we are not bypassing CSRF protection on any modifying endpoints.  Or, maybe we need to see if there's a way that this initial statistical call can be _delayed_ so that it doesn't run until after the initial CSRF token has been processed.

See our docs https://github.com/DSpace/RestContract/blob/main/csrf-tokens.md which say:
> All modifying requests (POST, PUT, PATCH, DELETE, etc) require that the client send a valid CSRF token in the X-XSRF-TOKEN header, otherwise a 403 Forbidden error will occur.

**Comment by bram-atmire** [comment]
@tdonohue the crux here is that the initial POST to the /csrf endpoint will result in a 403 response, by design. That 403 response includes the Set-Cookie header for the DSPACE-XSRF-COOKIE value, to guarantee that subsequent POST calls will use the correct matching one.

We are really convinced that this is a good solution to an important problem, in fact we are already using it in PROD.
You can check out how it works live on https://repository.upenn.edu/home

![Screenshot 2024-02-07 at 09 34 40](https://github.com/DSpace/DSpace/assets/958314/7cce1cf0-6585-4535-bf4f-b6d17893d171)

Hope this helps to clarify and to accept our solution!!!

**Comment by tdonohue** [comment]
@bram-atmire : feel free to create a PR for review. 

To clarify, my only objection to @benbosman 's earlier idea was that we use a `POST /csrf`.  I think it's more secure to use a `GET /csrf` and **hold back all modifying calls (POST/PATCH/DELETE) until that GET returns**.

Beyond that, the design sounds fine to me.  I'm just concerned about the approach of using POST to generate a CSRF token, as I believe that would _require bypassing at least part of the built in Spring protections._  If I'm wrong about that, then I'd love to be proven wrong.

**Comment by bram-atmire** [comment]
We would really welcome review here and hope to get this merged.

**Comment by tdonohue** [comment]
NOTE: The new CSRF code added in the attached PRs likely will be incompatible with Spring Security 6, which we are upgrading to in #9321 (as Spring Security 6 made some major changes in CSRF tokens).   So, we will need to analyze whether this problem still exists after upgrading to Spring Security 6...and, if so, what the best solution will be.  

One possible approach mentioned in the "Preparing for 6.0" Spring Security docs is to create a custom `GET /csrf` endpoint which will return the token whenever the client side application needs it.  See https://docs.spring.io/spring-security/reference/5.8/migration/servlet/exploits.html#_i_am_using_a_single_page_application_with_httpsessioncsrftokenrepository

Overall, the approach taken here may need more detailed analysis with regards to aligning with Spring Security.  The same approach may not work for both Spring Security 5 (which DSpace 7.x uses) and Spring Security 6 (which DSpace 8 will use).  I plan to look at this in more detail once 8.0 stabilizes, but I wanted to make sure others were aware that this concept may require more work.

**Comment by tdonohue** [comment]
This has been fixed on `main` by these PRs (which were inspired by / borrowed from Atmires work):
* #8713 added a new `GET /api/security/csrf` endpoint
* https://github.com/DSpace/dspace-angular/pull/2886 ensured that endpoint is used by DSpace UI to get a CSRF token when one has not yet been obtained.  This cherry-picked a commit from @artlowel in https://github.com/DSpace/dspace-angular/pull/2838 ... but updated it to work with the new backend code.

This fix has NOT yet been backported to `dspace-7_x`.  I haven't had time to do so, and it probably should be discussed first as it requires the creation of a new REST endpoint (`/api/security/csrf`).

**Comment by bram-atmire** [comment]
@tdonohue We think the decision to go with a GET /csrf endpoint really needs to be revised.

1. Because only POST requests should change state, get requests and responses should not include csrf tokens at all, in order to avoid token leaks
2. By giving back the tokens in a get call, bots are more likely to get them as they are more likely to perform get requests against endpoints vs post.

### External resource on this

See section "Disclosure of Token in URL" of
https://jcarpizo.github.io/owasp-info/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html

_The ideal solution is to only include the CSRF token in POST requests and modify server-side actions that have state changing affect to only respond to POST requests. This is in fact what the RFC 2616 requires for GET requests. If sensitive server-side actions are guaranteed to only ever respond to POST requests, then there is no need to include the token in GET requests._

Due to the fact that the GET endpoint became the community standard as of 7.6.2, we are now getting client requests to include support for this endpoint (for integration usecases) and we really don't want to add that.

Could you elaborate more on why you consider the GET approach more secure than the POST approach?

**Comment by tdonohue** [comment]
@bram-atmire : Let's create a different ticket for this discussion.  The bug described in this ticket has been fixed as of 7.6.2 / 8.0, I'd rather move this to a new ticket.  That way we don't get confusion over which changes were made in which release.

The primary reason I was in favor of `GET /csrf` is because that is the recommendations set forth by Spring Security. 

For example, here's Spring Security's recommendations for CSRF token generation for a single page application.  Notice it shows creating a `GET /csrf` endpoint, which should be used by the client side application:
https://docs.spring.io/spring-security/reference/5.8/migration/servlet/exploits.html#_i_am_using_a_single_page_application_with_httpsessioncsrftokenrepository

That said, these best practices do evolve over time. We can reanalyze and rethink this as necessary.

As for the resource you link to... I believe this text has been misunderstood:
> The ideal solution is to only include the CSRF token in POST requests and modify server-side actions that have state changing affect to only respond to POST requests.

I understand that to simply be saying that in *every POST request* you MUST also submit a CSRF Token for validation because those requests change state (as it goes on to say that you should not submit the CSRF Token in GET requests).  I do NOT believe that says that you should use `POST /csrf` to obtain or update the CSRF token.  So, I understand this statement as aligning with what we already do within DSpace (because we always submit a CSRF token with every POST request, and never with a GET request).

Regardless, let's move this to a different ticket to avoid confusing what features/updates have occurred in 7.6.2 / 8.0.  I'm not against deciding to change this, but I feel we are better off to **align with Spring Security**, since they are aligning with best practices for OWASP, etc.

Closing this ticket again as this is "fixed" in 7.6.2/8.0.  Please open a new ticket for further discussion
